solutions/platinum/cf-1254D.mdx

```diff
@@ -0,0 +1,151 @@
+---
+id: cf-1254D
+source: CF
+title: Tree Queries
+author: Rameez Parwez
+---
+
+[Official Editorial (C++)](https://codeforces.com/blog/entry/71594)
+
+## Explanation
+
+The idea is to use Heavy-Light Decomposition(HLD) for its efficiency in dividing the tree into chains and ensuring that paths between any two nodes pass through a limited number of chains, along with Binary Indexed Tree(BIT) to efficiently manage cumulative sums and range updates operations.
+
+At first, the DFS is performed starting from node $0$ to calculate the depth and subtree size. Afterwards, the decomposition is performed to transform the tree into a structure suitable for BIT operations.
+
+For a modification, we first check if vertex $v$ has some subtree. If it does, we want to add a tag to $v$ which increase the subtree of every light child $u$ of $v$ by $(n - x) * d$. When handling the query, we find its value on BIT and sum of all tags on its path to the root.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O} (N + Q \log(N))$
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MAX_N = 150001, MOD = 998244353;
+vector<int> adj[MAX_N], sum(MAX_N);
+vector<int> top(MAX_N), in(MAX_N), parent(MAX_N), depth(MAX_N), out(MAX_N),
+    sz(MAX_N);
+int T, n, q;
+
+vector<int> bit(MAX_N);
+void add(int x, int y) {
+	while (x <= MAX_N) {
+		bit[x] = (bit[x] + y) % MOD;
+		x += x & -x;
+	}
+}
+
+void add(int l, int r, int x) {
+	add(l + 1, x);
+	add(r + 1, MOD - x);
+}
+
+int get(int x) {
+	++x;
+	int res = 0;
+	while (x > 0) {
+		res = (res + bit[x]) % MOD;
+		x -= x & -x;
+	}
+	return res;
+}
+
+int power(int a, int b) {
+	int res = 1;
+	while (b) {
+		if (b & 1) { res = 1LL * res * a % MOD; }
+		a = 1LL * a * a % MOD;
+		b >>= 1;
+	}
+	return res;
+}
+
+// calculates depth and subtree size for each node
+void dfs_sz(int u) {
+	sz[u] = 1;
+	if (parent[u] != -1) {
+		adj[u].erase(find(adj[u].begin(), adj[u].end(), parent[u]));
```

Comment:
this feels unnecessary to me. why not just add like,

`if (v == parent[u]) { continue; }`

in the loop

Reply:
bcz sometimes removing the parent before the loop proves to be efficient in some cases.

Reply:
1. I don't think any other solution in the guide does this
2. vector .erase is O(size(vector)), so I feel like this isn't very efficient compared to just ignoring it lol

Reply:
if i change that then i will have to change the implementation in $modily$ function too where I am checking a condition in if statement and adding some values.

Reply:
if i change that i am messing up the solution since I am also adding some values to the final solution in $modify$ function mentioned below, where I am adding some values only for the vertex having some subtree.

Reply:
It's not giving TLE. The time constraint for this problem was 5s and my solution runs around ~1s. I would consider it good.

Reply:
> i mean this seems like it would tle unless you can give an explanation on why it wouldn't

ok it's not gonna TLE, because there are O(M) edges. however, it is kind of inefficient I feel.

Reply:
On the other hand, I fixed the solution without using the erase function. If you are insisting on that I am ready to change it.

Reply:
@TheGamingMousse code updated. could you take a look at it?

Reply:
> @TheGamingMousse code updated. could you take a look at it?

I'm confused, aren't u still removing the parent directly?

Reply:
@TheGamingMousse I think you are looking at the old code. Check the new commit.

Reply:
![image](https://github.com/user-attachments/assets/1ded18d7-9c8d-4565-8b33-bc4c65943e6c)

doesn't this still remove the parent from the adj list? lol

Reply:
This is the solution of Subtree and Paths from HackerRank. 
At first, you were talking about Tree Queries from CF.

I used the same concept on both of the problems, but since I was too much indulged in CF, that i forgot about this.

btw when I submitted the problem with and without removing the parent I got somewhere around the same time constraint for both the cases, there wasn't much difference.

Reply:
idk it's just a bit odd and isn't really used in other guide code

also like, you have the `if (v != p)` check anyways, so what's the point

Reply:
write now if i am making any changes with that then it is resulting in segmentation fault error.
I will try to figure it out later and update it.

Reply:
wacky stuff

Reply:
@TheGamingMousse updated the code.

Reply:
> wacky stuff

I always find it hard to deal with SIGSEGV errors.